app/models/bus_stop.rb

```diff
@@ -250,8 +250,16 @@ def name_with_id
     "#{name} (#{hastus_id})"
   end
 
-  def self.strip_id_from_name(name_with_id)
-    name_with_id[/[^(]+/].strip
+  def self.find_by_name_search(search_query)
+    id_string_matches = search_query.match(/ +\(\d+\)$/)
+    if id_string_matches.present?
+      extracted_id_string = id_string_matches[0]
+      extracted_hastus_id = extracted_id_string[/\d+/].to_i
+      stop_by_id = BusStop.find_by(hastus_id: extracted_hastus_id)
+      return stop_by_id if stop_by_id.present?
+      search_query = search_query.gsub extracted_id_string, ''
+    end
+    BusStop.find_by(name: search_query)
```

Comment:
Finished code golfing:
```suggestion
      hastus_string = search_query.match(/ +\(\d+\)$/).to_s 
      hastus_id = hastus_string.match(/\d+/).to_s
      name = search_query.gsub hastus_string, ''
      BusStop.find_by(hastus_id: hastus_id) || BusStop.find_by(name: name)
```
One thing I should mention is that you don't have to turn strings into integers to search by IDs. Rails is nice and knows what you mean (we do it in the params all the time if you recall; they're always strings).
Like, `User.find "6"` will work just as well as `User.find 6`